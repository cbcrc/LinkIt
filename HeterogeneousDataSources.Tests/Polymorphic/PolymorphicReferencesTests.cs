using System.Collections.Generic;
using ApprovalTests.Reporters;
using HeterogeneousDataSources.ConfigBuilders;
using HeterogeneousDataSources.LinkedSources;
using HeterogeneousDataSources.Protocols;
using HeterogeneousDataSources.Tests.Shared;
using NUnit.Framework;
using RC.Testing;

namespace HeterogeneousDataSources.Tests.Polymorphic {
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class PolymorphicReferencesTests {
        private FakeReferenceLoader<Model, string> _fakeReferenceLoader;
        private LoadLinkProtocol _sut;

        [SetUp]
        public void SetUp() {
            var loadLinkProtocolBuilder = new LoadLinkProtocolBuilder();

            loadLinkProtocolBuilder.For<LinkedSource>()
                .PolymorphicLoadLinkForList(
                    linkedSource => linkedSource.Model.Target,
                    linkedSource => linkedSource.Target,
                    link => link.Type,
                    includes => includes
                        .Include<Image>().AsReference(
                            "image",
                            link=>link.Id
                        )
                        .Include<Person>().AsReference(
                            "person",
                            link => link.Id
                        )
                );

            _fakeReferenceLoader =
                new FakeReferenceLoader<Model, string>(reference => reference.Id);
            _sut = loadLinkProtocolBuilder.Build(_fakeReferenceLoader);
        }

        [Test]
        public void LoadLink_PolymorphicReferenceWithImage() {
            _fakeReferenceLoader.FixValue(
                new Model {
                    Id = "1",
                    Target = new List<PolymorphicReference>
                    {
                        new PolymorphicReference {
                            Type = "person",
                            Id = "a"
                        },
                        new PolymorphicReference{
                            Type = "image",
                            Id = "a"
                        }
                    }
                }
            );

            var actual = _sut.LoadLink<LinkedSource>().ById("1");

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        public class LinkedSource : ILinkedSource<Model> {
            public Model Model { get; set; }
            public List<object> Target { get; set; }
        }

        public class Model{
            public string Id { get; set; }
            public List<PolymorphicReference> Target { get; set; }
        }

        //stle: should be shared
        public class PolymorphicReference {
            public string Type { get; set; }
            public string Id { get; set; }
        }
    }
}
