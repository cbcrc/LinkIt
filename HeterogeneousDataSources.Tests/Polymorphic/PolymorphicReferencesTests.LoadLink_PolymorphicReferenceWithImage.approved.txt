{
  "model": {
    "id": "1",
    "target": [
      {
        "type": "person",
        "id": "a"
      },
      {
        "type": "image",
        "id": "a"
      }
    ]
  },
  "target": [
    {
      "$type": "LinkIt.Tests.Shared.Person, LinkIt.Tests",
      "id": "a",
      "name": "name-a",
      "summaryImageId": "person-img-a"
    },
    {
      "$type": "LinkIt.Tests.Shared.Image, LinkIt.Tests",
      "id": "a",
      "alt": "alt-a"
    }
  ]
}